stout/composed-loop.h

```diff
@@ -0,0 +1,154 @@
+#pragma once
+
+#include <optional>
+
+#include "stout/stream.h"
+#include "stout/terminal.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace stout {
+namespace eventuals {
+
+////////////////////////////////////////////////////////////////////////
+
+using namespace detail;
+
+struct _ComposedLoopRouter {
+  template <typename PrevK_, typename Arg_>
+  struct Continuation {
+    void Start(TypeErasedStream& stream) {
+      stream.Next();
+    }
+
+    template <typename... Args>
+    void Body(Args&&... args) {
+      stored_k_->k_.Body(std::forward<Args>(args)...);
+    }
+
+    void Ended() {
+      stored_k_->Done();
+    }
+
+    PrevK_* stored_k_;
+  };
+
+  template <typename ComposedK>
+  struct Composable {
+    template <typename Arg>
+    using ValueFrom = Arg;
+
+    template <typename Arg>
+    auto k() && {
+      return Continuation<ComposedK, Arg>{std::move(composed_k)};
+    }
+
+    ComposedK* composed_k;
+  };
+};
+
+template <typename K>
+auto ComposedLoopRouter(K* k) {
+  return _ComposedLoopRouter::Composable<K>{k};
+}
+
+////////////////////////////////////////////////////////////////////////
+
+struct _ComposedLoop {
+  template <typename K_, typename F_, typename Arg_>
+  struct Continuation : public TypeErasedStream {
+    // NOTE: explicit constructor because inheriting 'TypeErasedStream'.
+    Continuation(K_ k, F_ f)
+      : k_(std::move(k)),
+        f_(std::move(f)) {}
+
+    void Start(TypeErasedStream& stream) {
+      stream_ = &stream;
+      k_.Start(*this);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    template <typename... Args>
+    void Body(Args&&... args) {
+      // Executes when Body is called from the above-level stream.
+      if (innerStream_ == nullptr) {
+        inner_.emplace(Build(operator|(f_(std::forward<Args>(args)...), ComposedLoopRouter(this))));
+        innerStream_ = &inner_.value();
+        inner_.value().Start();
+      } else {
+        // Executes when Body is called from inner stream Next function.
+        k_.Body(std::forward<Args>(args)...);
+      }
+    }
+
+    void Ended() {
+      // When the topmost loop have no more values there no inner stream
+      // initialization, so it can indicate the loop end.
+      if (innerStream_ == nullptr) {
+        k_.Ended();
+        return;
+      }
+
+      // Trying to get new value from the loop above.
+      innerStream_ = nullptr;
+      stream_->Next();
+    }
+
+    void Next() override {
+      if (innerStream_ == nullptr) {
+        stream_->Next();
+        return;
+      }
+      innerStream_->Next();
+    }
+
+    void Done() override {
+      Ended();
+    }
+
+    using Left_ = typename std::invoke_result<F_, int>::type;
```

Comment:
Will `F_` always be invoked with an `int`?

Reply:
whoops, my fault